Fix autocast for non-strict export

yushangdi opened the pull request:
Summary: 

add testing for autocast and set_grad nodes for export_for_training. In export_for_training, we do not wrap the autocast and set_grad node in to HOP, but we should still have the set_grad_enabled/autocast nodes.

add support for autocast in non-strict export. Previously, `_enter_autocast` and `_exit_autocast` nodes don't show up in the export graph when we use `strict=False`.

- In autocast's enter and exit function, we dispatch to `PreDispatchTorchFunctionMode.__torch_function__`.
 if we have PreDispatchTorchFunctionMode in our function_mode_stack, the call stack looks like below. This is mostly the same call stack as strict mode, except strict mode enters [here](https://www.internalfb.com/code/fbsource/[0d4f1135cacdb26c6e01d5dce1ce52a15d61ee48]/xplat/caffe2/torch/_dynamo/variables/ctx_manager.py?lines=806).
```
- torch.amp.autocast.__enter__()'s torch.overrides.handle_torch_function
- torch.fx.experimental.proxy_tensor.TorchFunctionMetadataMode.__torch_function__
- torch.amp._enter_autocast()'s torch.overrides.handle_torch_function
- PreDispatchTorchFunctionMode.__torch_function__
```
- in `PreDispatchTorchFunctionMode.__torch_function__`, we create the autocast nodes.
- to match the strict mode behavior, we let the input node to the `_exist_autocast` node be the corresponding `_enter_autocast` node. This requires us to maintain a stack in `PreDispatchTorchFunctionMode`.

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_export_with_autocast
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_export_with_set_grad
```

Differential Revision: D64016023




cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

bdhirsh commented on torch/amp/autocast_mode.py:
```diff
@@ -355,10 +355,25 @@ def __enter__(self):
 
+        if torch._C._is_torch_function_mode_enabled():
+            return torch.overrides.handle_torch_function(
+                torch.amp._enter_autocast,
```
instead of overring the `_enter/exit_autocast` calls in python here, can we just override the single `_set_autocast_enabled(bool)` function from C++, similar to what we do for `_set_autograd_enabled` here? https://github.com/pytorch/pytorch/blob/9b2e453e246ea44f2071dfdac86ec4b0037a51a5/torch/csrc/autograd/init.cpp#L877

bdhirsh replied:
cc @zou3519 / @albanD in case you have any thoughts/preference. The context is that we need a __torch_function__ override for autocast in order to preserve autocast for non-strict, pre-dispatch export.

And we have two options:

(1) override all 3 mutable autocast API's:
```
torch.set_autocast_enabled
torch.set_autocast_dtype
torch.set_autocast_cache_enabled
```

(2) override the enter/exit API's:
```
torch.amp._enter_autocast
torch.amp._exit_autocast
```

(1) will be a little more in-line with how we override `torch.set_grad_enabled` today (when desugaring `no_grad/enable_grad`). Although it's a bit silly because there are 3 individual mutable autocast API's we'll need to expose to `__torch_function__` instead of 2. If we do (1),  @yushangdi also pointed out that dynamo currently does (2) today, so we'll probably want to fix up dynamo to desugar to (1) for consistency between strict/nonstrict

(2) is what this PR does today.

(I think either is fine, although if we choose one now and want to switch later, there is a bit of BC risk for `__torch_function__` authors)

yushangdi replied:
Summarizing chat content:

We could potentially override the C++ API of the following, 

torch.set_autocast_enabled
torch.set_autocast_dtype
torch.set_autocast_cache_enabled

but then it would make strict and non-strict export behavior different. Specifically, in the strict mode produce a single node torch.amp.autocast_mode._enter_autocast (or exit) here: https://github.com/pytorch/pytorch/blob/487873f7cafeb0fd390eaefe40496b804bceabbd/torch/_dynamo/variables/ctx_manager.py#L808

If we override the C++ API, then we would produce three nodes instead. It would also make subsequent pattern matching more complicated.

I guess autocast is a little different from grad_enbaled in that set_autocast_enabled itself is not equivalent to _enter_autocast. We also need a few other operations.

yushangdi replied:
minor comment: for option 1, I think we might need to override two additional APIs: 

torch.autocast_increment_nesting()
torch.clear_autocast_cache()

zou3519 replied:
torch_function is about overriding the public PyTorch API, so my take is we shouldn't do (2)

yushangdi replied:
> set_autocast_enabled

@zou3519 @bdhirsh  I found that `set_autocast_enabled` is a public function that cannot be overridden by ``__torch_function__``. The docstring says this is because none of the arguments of these functions are tensors or tensor-likes.

Does this matter to what we are dicussing?

https://github.com/pytorch/pytorch/blob/main/torch/overrides.py#L100C12-L100C81

https://github.com/pytorch/pytorch/blob/main/torch/overrides.py#L269

yushangdi replied:
Do you think we can do something like below (might be hacky):

We only go to `handle_torch_function` if we have PreDispatchTorchFunctionMode on the functional mode stack and we only dispatch to PreDispatchTorchFunctionMode's torch_function, not any other FunctionMode.

```python
	
        if torch._C._is_torch_function_mode_enabled():
            from torch.fx.experimental.proxy_tensor import PreDispatchTorchFunctionMode

            mode_stack = torch.overrides._get_current_function_mode_stack()
            if any(
                isinstance(mode, PreDispatchTorchFunctionMode) for mode in mode_stack
            ):
                return torch.overrides.handle_torch_function(
                    torch.amp._exit_autocast, [], self
                )
```

bdhirsh replied:
> I found that set_autocast_enabled is a public function that cannot be overridden by __torch_function__ (similarly with the other autocast functions). The docstring says this is because none of the arguments of these functions are tensors or tensor-likes.

This is true for torch-function based tensor-subclasses, but not for modes (which can intercept factory-function-style torch ops). If you look at how we handle `set_autograd_enabled` ([link](https://github.com/pytorch/pytorch/blob/9b2e453e246ea44f2071dfdac86ec4b0037a51a5/torch/csrc/autograd/init.cpp#L875)) - we check that a `TorchFunctionMode` is enabled before dispatching to `handle_torch_function()`

zou3519 replied:
Option 1 with a mode sounds like the most straighforward thing to do. An alternative is that we do option 2, but we expose those APIs to users (or figure out what we do want to expose to users).

yushangdi replied:
> Option 1 with a mode sounds like the most straighforward thing to do. An alternative is that we do option 2, but we expose those APIs to users (or figure out what we do want to expose to users).

I changed to Option 1 with a mode now. Thanks!

bdhirsh reviewed (approved):
preemptive stamp

yushangdi committed:
autocast in non-strict mode (#137495)

Summary:

add testing for autocast and set_grad nodes for export_for_training. In export_for_training, we do not wrap the autocast and set_grad node in to HOP, but we should still have the set_grad_enabled/autocast nodes.

add support for autocast in non-strict export. Previously, `_enter_autocast` and `_exit_autocast` nodes don't show up in the export graph when we use `strict=False`.

- In autocast's enter and exit function, we dispatch to `PreDispatchTorchFunctionMode.__torch_function__`.
 if we have PreDispatchTorchFunctionMode in our function_mode_stack, the call stack looks like below. This is mostly the same call stack as strict mode, except strict mode enters [here](https://www.internalfb.com/code/fbsource/[0d4f1135cacdb26c6e01d5dce1ce52a15d61ee48]/xplat/caffe2/torch/_dynamo/variables/ctx_manager.py?lines=806).
```
- torch.amp.autocast.__enter__()'s torch.overrides.handle_torch_function
- torch.fx.experimental.proxy_tensor.TorchFunctionMetadataMode.__torch_function__
- torch.amp._enter_autocast()'s torch.overrides.handle_torch_function
- PreDispatchTorchFunctionMode.__torch_function__
```
- in `PreDispatchTorchFunctionMode.__torch_function__`, we create the autocast nodes.
- to match the strict mode behavior, we let the input node to the `_exist_autocast` node be the corresponding `_enter_autocast` node. This requires us to maintain a stack in `PreDispatchTorchFunctionMode`.

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_export_with_autocast
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:test_export  -- -r  test_export_with_set_grad
```

Reviewed By: bdhirsh

Differential Revision: D64016023
